### PR TITLE
Improve scarecrow

### PR DESCRIPTION
**Pull request by ShadowProtocol**
Implements #1855 

# Changes proposed in this pull request:
- Scarecrow (Field) now physically takes up two blocks
- Scarecrow GUI is accessible from either half of scarecrow 
- Fixed a glitch where dragging a seed from the field slot would duplicate the seed
    due to the wrong inventory handler
- Modified the scarecrow held item to conform slightly closer to the art style of similar held items

Please review.

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java [hunk 1]
```diff
@@ -119,87 +117,16 @@
     /**
      * Inventory of the field.
      */
-    private final IItemHandlerModifiable inventory = new IItemHandlerModifiable() {
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java:122]: As far as I know we should not be using IInventory anymore and instead use IItemhandlers.

File: src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java [hunk 2]
```diff
@@ -135,17 +167,27 @@ public void onExplosionDestroy(final World worldIn, final BlockPos pos, final Ex
     @Override
     public void onBlockHarvested(final World worldIn, @NotNull final BlockPos pos, final BlockState state, @NotNull final PlayerEntity player)
     {
+        DoubleBlockHalf half = state.get(HALF);
+        BlockPos otherpos = half == DoubleBlockHalf.LOWER? pos.up() : pos.down();
+        BlockState otherstate = worldIn.getBlockState(otherpos);
+
+        // just double check the other block is also the scarecrow and not the same half,
+        // then destroy it (make it air)
+        if (otherstate.getBlock() == this && otherstate.get(HALF) != half) {
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java]: formatting
{

}

(All over the PR)

File: src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java [hunk 3]
```diff
@@ -101,8 +114,23 @@ public ActionResultType onBlockActivated(
...
+        @NotNull final Direction Direction = (context.getPlayer() == null) ? NORTH : fromAngle(context.getPlayer().rotationYaw + 180);
+
+        if (context.getPos().getY() < 255 && context.getWorld().getBlockState(context.getPos().up()).isReplaceable(context)) {
+            return this.getDefaultState().with(FACING, Direction).with(HALF, DoubleBlockHalf.LOWER);
+        } else {
+            return null;
+        }
+    }
+
+    public boolean isValidPosition(BlockState state, IWorldReader worldIn, BlockPos pos) {
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java]: @Override

File: src/api/java/com/minecolonies/api/inventory/container/ContainerField.java [hunk 4]
```diff
@@ -87,28 +97,28 @@ public ContainerField(final int windowId, final PlayerInventory playerInventory,
...
+        transfer = slot.getStack();
+
+        if (index == 0)
         {
-            final int playerIndex = slotIndex < MAX_INVENTORY_INDEX ? (slotIndex + INVENTORY_BAR_SIZE) : (slotIndex - MAX_INVENTORY_INDEX);
-            if (playerIn.inventory.getStackInSlot(playerIndex) != ItemStackUtils.EMPTY)
+            if (!mergeItemStack(transfer, 1, 37, true)) {
+                return ItemStack.EMPTY;
+            }
+        } else {
```
Comment (Asherslab) [inline src/api/java/com/minecolonies/api/inventory/container/ContainerField.java]: formatting

File: src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java [hunk 5]
```diff
@@ -87,22 +95,49 @@ public ActionResultType onBlockActivated(
...
+            DoubleBlockHalf half = state.get(HALF);
+            final TileEntity entity = worldIn.getTileEntity(
+                half == DoubleBlockHalf.UPPER? pos.down() : pos
+            );
+
             if (entity instanceof ScarecrowTileEntity)
             {
                 NetworkHooks.openGui((ServerPlayerEntity) player, (ScarecrowTileEntity) entity, pos);
-                return ActionResultType.SUCCESS;
+            } else {
```
Comment (Asherslab) [inline src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java]: formatting

File: src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java [hunk 6]
```diff
@@ -87,22 +95,49 @@ public ActionResultType onBlockActivated(
...
     public BlockState getStateForPlacement(final BlockItemUseContext context)
     {
-        @NotNull final Direction Direction = (context.getPlayer() == null) ? NORTH : fromAngle(context.getPlayer().rotationYaw);
-        return this.getDefaultState().with(FACING, Direction);
+        @NotNull final Direction Direction = (context.getPlayer() == null) ? NORTH : fromAngle(context.getPlayer().rotationYaw + 180);
+
+        if (context.getPos().getY() < 255 && context.getWorld().getBlockState(context.getPos().up()).isReplaceable(context))
+        {
+            return this.getDefaultState().with(FACING, Direction).with(HALF, DoubleBlockHalf.LOWER);
+        } else {
```
Comment (Asherslab) [inline src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java]: formatting

File: src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java [hunk 7]
```diff
@@ -87,22 +95,49 @@ public ActionResultType onBlockActivated(
...
+
+    @Override
+    public boolean isValidPosition(BlockState state, IWorldReader worldIn, BlockPos pos)
+    {
+        BlockPos blockpos = pos.down();
+        BlockState blockstate = worldIn.getBlockState(blockpos);
+        if (state.get(HALF) == DoubleBlockHalf.LOWER)
+        {
+            return blockstate.isSolidSide(worldIn, blockpos, Direction.UP);
+        } else {
```
Comment (Asherslab) [inline src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java]: formatting

File: src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java [hunk 8]
```diff
@@ -135,17 +176,28 @@ public void onExplosionDestroy(final World worldIn, final BlockPos pos, final Ex
     @Override
     public void onBlockHarvested(final World worldIn, @NotNull final BlockPos pos, final BlockState state, @NotNull final PlayerEntity player)
     {
+        DoubleBlockHalf half = state.get(HALF);
+        BlockPos otherpos = half == DoubleBlockHalf.LOWER? pos.up() : pos.down();
```
Comment (Asherslab) [inline src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java]: formatting (space between LOWER and `?` in ternaries)

File: src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java [hunk 9]
```diff
@@ -135,17 +176,28 @@ public void onExplosionDestroy(final World worldIn, final BlockPos pos, final Ex
...
+        {
+            worldIn.setBlockState(otherpos, Blocks.AIR.getDefaultState(), 35);
+        }
+
         notifyColonyAboutDestruction(worldIn, pos);
         super.onBlockHarvested(worldIn, pos, state, player);
     }
-
+/*
     @Override
```
Comment (Asherslab) [inline src/main/java/com/minecolonies/coremod/blocks/BlockScarecrow.java]: please remove code don't comment it out

File: src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java (hunk 1 above)
Comment (ShadowProtocol) [reply src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java:122]: Ah, true true. There was still one of the functions causing a glitch in the implementation that was there before. I believe it is fixed now. Not quite sure which part was causing it still.

File: src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java [hunk 10]
```diff
@@ -578,7 +518,7 @@ public void read(final CompoundNBT compound)
             final ItemStack stack = ItemStack.read(inventoryCompound);
             if (ItemStackUtils.getSize(stack) <= 0)
             {
-                inventory.setStackInSlot(i, ItemStackUtils.EMPTY);
+                inventory.setStackInSlot(i, ItemStack.EMPTY);
```
Comment (Asherslab) [inline src/main/java/com/minecolonies/coremod/tileentities/ScarecrowTileEntity.java]: usage of `ItemStackUtils` version is preferred for easier future porting (reason it was made in the first place)

**Review by Asherslab — DISMISSED** [review]
looks good other than some cleanup

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/ldtteam/minecolonies/pull-request/5241/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

**Review by someaddons — APPROVED** [review]

**Comment by Strohtaler** [comment]
Great Job, working fine! and JIT, I was about to create an issue about the dragging seeds problem [: -)

**Pull request merged**